File: app/src/main/java/kr/ac/hansung/demap/ui/main/MainActivity.java
package kr.ac.hansung.demap.ui.main;

import android.Manifest;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.graphics.PointF;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.Window;
import android.widget.SearchView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.view.GravityCompat;
import androidx.drawerlayout.widget.DrawerLayout;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.navigation.NavigationView;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.naver.maps.geometry.LatLng;
import com.naver.maps.map.CameraAnimation;
import com.naver.maps.map.CameraUpdate;
import com.naver.maps.map.MapView;
import com.naver.maps.map.NaverMap;
import com.naver.maps.map.OnMapReadyCallback;
import com.naver.maps.map.overlay.LocationOverlay;
import com.naver.maps.map.overlay.Marker;
import com.naver.maps.map.overlay.OverlayImage;
import com.naver.maps.map.util.FusedLocationSource;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.List;

import kr.ac.hansung.demap.CreateFolderActivity;
import kr.ac.hansung.demap.CustomerCenterActivity;
import kr.ac.hansung.demap.ui.searchfolder.FolderListActivity;
import kr.ac.hansung.demap.ui.myfolderlist.MyfolderViewActivity;
import kr.ac.hansung.demap.ui.notice.NoticeActivity;
import kr.ac.hansung.demap.R;
import kr.ac.hansung.demap.ui.searchplace.SearchNaverActivity;
import kr.ac.hansung.demap.ui.setting.SettingsActivity;
import kr.ac.hansung.demap.ui.hotPlace.HotPlaceActivity;
import kr.ac.hansung.demap.ui.login.LoginActivity;
import kr.ac.hansung.demap.ui.nickname.NickNameActivity;

public class MainActivity extends AppCompatActivity implements OnMapReadyCallback, NavigationView.OnNavigationItemSelectedListener, BottomNavigationView.OnNavigationItemSelectedListener, View.OnClickListener {

    private FirebaseAuth auth = FirebaseAuth.getInstance();
    private FirebaseFirestore firestore = FirebaseFirestore.getInstance();

    private Intent settingsIntent;
    private Intent myFolderIntent;

    private String uid;
    private String nickname = "";
    private TextView tv_nickname;
    private TextView tv_email;

    private NavigationView navigationView;
    private BottomNavigationView bottomNavigationView;

    private DrawerLayout drawerLayout;

    private MapView mapView;
//    private LocationButtonView locationButtonView;

    private SearchView searchView;

    // FusedLocationSource (Google)
    private static final int LOCATION_PERMISSION_REQUEST_CODE = 1000;
    private FusedLocationSource locationSource;
    private static Geocoder geocoder;


    private static double latitude;
    private static double altitude;
    private static double longitude;

    private static double latitude_togo = 37.5666103;
    private static double longitude_togo = 126.9783882;
    private static List<Address> location_name;
    private static List<Address> location_name_togo;

    private FloatingActionButton fab_now;
    private FloatingActionButton fab_navi;

    static NaverMap naverMap_keep;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);


        // 플로팅 버튼 생성
        fab_now = (FloatingActionButton) findViewById(R.id.fab_now_point);
        fab_now.setOnClickListener(this);
        fab_navi = (FloatingActionButton) findViewById(R.id.fab_navi);
        fab_navi.setOnClickListener(this);

        geocoder = new Geocoder(this);

        Toolbar toolbar = findViewById(R.id.toolbar_main);
        setSupportActionBar(toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true); // 드로어를 꺼낼 홈 버튼 활성화
        getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_menu_black_24dp); // 홈버튼 이미지 변경
        getSupportActionBar().setDisplayShowTitleEnabled(false); // 툴바에 타이틀 안보이게
        getSupportActionBar().setBackgroundDrawable(getDrawable(R.color.colorWhite)); // 배경색

        Window window = getWindow();
        window.setStatusBarColor(getResources().getColor(R.color.colorWhite));

        settingsIntent = new Intent(this, SettingsActivity.class);
        myFolderIntent = new Intent(this, MyfolderViewActivity.class);

        drawerLayout = findViewById(R.id.drawerlayout_main);

/**  내 위치 리스너 **/
        // GPS 연동을 위한 권한 체크
        /*
        if (Build.VERSION.SDK_INT >= 23 &&
                ContextCompat.checkSelfPermission(getApplicationContext(),
                        android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((Activity)this,
                    new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION},
                    0);
        }
        */


        // 로그인한 유저 닉네임 받아오기
        if (auth.getCurrentUser() != null) {
            uid = auth.getCurrentUser().getUid();
            settingsIntent.putExtra("uid", uid);
            firestore.collection("users").document(uid).get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                @Override
                public void onSuccess(DocumentSnapshot documentSnapshot) {
                    if (documentSnapshot.exists()) {
                        nickname = documentSnapshot.get("nickName").toString();
                        tv_nickname = findViewById(R.id.tv_main_nickname);
                        tv_nickname.setText(nickname);
                        tv_email = findViewById(R.id.tv_main_email);
                        tv_email.setText(auth.getCurrentUser().getEmail());

                        settingsIntent.putExtra("nickname", nickname);
                        myFolderIntent.putExtra("nickname", nickname);
                    }
                    else {
                        Intent intent = new Intent(getApplicationContext(), NickNameActivity.class);
                        startActivity(intent);
                    }
                }
            });
        }

        navigationView = findViewById(R.id.main_nav);
        navigationView.setNavigationItemSelectedListener(this);

        bottomNavigationView = findViewById(R.id.bottom_nav);
        bottomNavigationView.setOnNavigationItemSelectedListener(this);

//        MapFragment mapFragment = (MapFragment) getSupportFragmentManager().findFragmentById(R.id.map_fragment);

        mapView = findViewById(R.id.navermap_map_view);
        mapView.onCreate(savedInstanceState);

        naverMapBasicSettings();

        searchView = findViewById(R.id.sv_searchPlace);
        searchView.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.d("log", "검색 버튼 눌림");
                searchView.clearFocus();
                searchPlace(query); // 검색화면
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return false;
            }
        });

    }

    public void naverMapBasicSettings() {
        mapView.getMapAsync(this);
        //내위치 버튼
//        locationButtonView = findViewById(R.id.locationbuttonview);
        // 내위치 찾기 위한 source
        // 오류나서 지움
//        locationSource = new FusedLocationSource(this, LOCATION_PERMISSION_REQUEST_CODE);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
//        if (locationSource.onRequestPermissionsResult(requestCode, permissions, grantResults)) {
//            return;
//        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private void nowMyPoint(NaverMap naverMap) throws IOException {
        /**  내 위치 리스너 **/
        // GPS 연동을 위한 권한 체크
        if (Build.VERSION.SDK_INT >= 23 &&
                ContextCompat.checkSelfPermission(getApplicationContext(),
                        android.Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions((Activity)this,
                    new String[]{android.Manifest.permission.ACCESS_FINE_LOCATION},
                    0);
        } else {
            // 내위치 검색
            LocationManager lm = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
            Location location = lm.getLastKnownLocation(LocationManager.GPS_PROVIDER);
            if(location==null){
                //gps를 이용한 좌표조회 실패시 network로 위치 조회
                location = lm.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
            }
            String provider = location.getProvider();
            double now_longitude = location.getLongitude();
            double now_latitude = location.getLatitude();
            double now_altitude = location.getAltitude();
            location_name = geocoder.getFromLocation(now_latitude,now_longitude,1);

            latitude = now_latitude;
            altitude = now_altitude;
            longitude = now_longitude;

            lm.requestLocationUpdates(LocationManager.GPS_PROVIDER,
                    3000,
                    1,
                    gpsLocationListener);
            lm.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,
                    3000,
                    1,
                    gpsLocationListener);

            // 카메라 위치 변경
            CameraUpdate cameraUpdate = CameraUpdate.scrollTo(new LatLng(now_latitude, now_longitude)).animate(CameraAnimation.Easing);
            naverMap.moveCamera(cameraUpdate);

            // 현재 위치 오버레이
            LocationOverlay locationOverlay = naverMap.getLocationOverlay();
            locationOverlay.setVisible(true);

            locationOverlay.setPosition(new LatLng(now_latitude, now_longitude));
            locationOverlay.setSubIconWidth(80);
            locationOverlay.setSubIconHeight(40);
            locationOverlay.setSubAnchor(new PointF(0.5f, 1));
        }
    }

    @Override
    public void onMapReady(@NonNull NaverMap naverMap) {

        naverMap_keep = naverMap;

        try {
            nowMyPoint(naverMap);
        } catch (IOException e) {
            System.out.println("현재 위치 가져오기 에러");
        }

        Marker marker = new Marker();
        marker.setPosition(new LatLng(37.5666103, 126.9783882));
       // marker.setPosition(new LatLng(latitude, longitude));
        System.out.println("내 현재위치 : " + latitude +", " + longitude);
//        marker.setIcon(OverlayImage.fromResource(R.drawable.icon_place_marker_for_map));
        marker.setIcon(OverlayImage.fromResource(R.drawable.icon_place_marker));
        marker.setWidth(100);
        marker.setHeight(120);
        marker.setMap(naverMap);

        try {
            location_name_togo = geocoder.getFromLocation(latitude_togo,longitude_togo,1);
        } catch (IOException e) {
            System.out.println("도착지 주소 가져오기 실패");
        }

        // 지도 아무데나 눌렀을 때
        naverMap.setOnMapClickListener(new NaverMap.OnMapClickListener() {
            @Override
            public void onMapClick(@NonNull PointF pointF, @NonNull LatLng latLng) {
                //Log.d("log", "마커이동"+latLng.toString());


                LatLng now = latLng;
                if (now != marker.getPosition()) {
                    System.out.println("마커이동"+latLng.toString());
                    marker.setMap(null);
                    marker.setPosition(latLng);
                    marker.setMap(naverMap);

                    latitude_togo = latLng.latitude;
                    longitude_togo = latLng.longitude;
                    try {
                        location_name_togo = geocoder.getFromLocation(latitude_togo,longitude_togo,1);
                    } catch (IOException e) {
                        System.out.println("도착지 주소 가져오기 실패");
                    }

                }



            }
        });

    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.fab_now_point:
                try {
                    nowMyPoint(naverMap_keep);
                } catch (IOException e) {
                    System.out.println("현재 위치 가져오기 에러");
                }
                break;
            case R.id.fab_navi:
                String sname = "현재위치";
                String dname = "도착위치";

                try {
                    String loc_name_str = location_name.toString();
                    String data[] = loc_name_str.split("\"");
                    System.out.println("현재위치주소"+data[1]);
                    String loc_name_str2 = location_name_togo.toString();
                    String data2[] = loc_name_str2.split("\"");
                    System.out.println("도착위치주소"+data[1]);
                    sname = URLEncoder.encode(data[1], "UTF-8");
                    dname = URLEncoder.encode(data2[1], "UTF-8");

                } catch (UnsupportedEncodingException e) {
                    System.out.println("현재위치 utf-8 인코딩 에러");
                }

                String url = "nmap://route/public?slat="+latitude+"&slng="+longitude+
                        "&sname="+sname+"&dlat="+latitude_togo+"&dlng="+longitude_togo+
                "&dname="+dname+"&appname=kr.ac.hansung.demap";

                Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
                intent.addCategory(Intent.CATEGORY_BROWSABLE);

                List<ResolveInfo> list = getPackageManager().queryIntentActivities(intent, PackageManager.MATCH_DEFAULT_ONLY);
                if (list == null || list.isEmpty()) {
                    view.getContext().startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=com.nhn.android.nmap")));
                } else {
                    view.getContext().startActivity(intent);
                }
                break;
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        mapView.onStart();
    }

    @Override
    protected void onResume() {
        super.onResume();
        mapView.onResume();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mapView.onPause();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        mapView.onSaveInstanceState(outState);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mapView.onStop();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }

    public void searchPlace(String searchText) {
        Intent intent = new Intent(this, SearchNaverActivity.class);
        intent.putExtra("searchText", searchText);
        startActivity(intent);
    }

    public void createFolder() {
        Intent intent = new Intent(this, CreateFolderActivity.class);
        intent.putExtra("folder_edit_flag", "create");
        intent.putExtra("fromMain", true);
        startActivity(intent);
    }

    public void viewFolderList() {
        Intent intent = new Intent(this, FolderListActivity.class);
        startActivity(intent);
    }

    public void viewMyFolderList() {
        startActivity(myFolderIntent);
    }

    public void viewSearchHotPlace() {
        Intent intent = new Intent(this, HotPlaceActivity.class);
//        intent.putExtra("search_hint", "맛집 검색");
        startActivity(intent);
    }

    public void notice() {
        Intent intent = new Intent(this, NoticeActivity.class);
        startActivity(intent);
    }

    public void service() {
        Intent intent = new Intent(this, CustomerCenterActivity.class);
        startActivity(intent);
    }

    public void settings() {
        startActivity(settingsIntent);
    }

    public void logout() {
        auth.signOut();
        Intent intent = new Intent(this, LoginActivity.class);
        startActivity(intent);
        finish();
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home){
            drawerLayout.openDrawer(GravityCompat.START); // 네비게이션 드로어 열기
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.search_folder_menu: viewFolderList(); break; // 폴더 검색
            case R.id.my_folder_menu: viewMyFolderList(); break; // 마이 폴더
            case R.id.hotplace_menu: viewSearchHotPlace(); break; // 핫플레이스
            case R.id.history_menu: notice(); break;
            case R.id.setting_menu: settings(); break;
            case R.id.service_menu: service(); break;
            case R.id.logout_menu: logout(); break;

            //바텀
            case R.id.home_nav: break;
            case R.id.history_bottom_nav: notice(); break;
            case R.id.search_folder_bottom_nav: viewFolderList(); break; // 폴더 검색
            case R.id.my_folder_bottom_nav: createFolder(); break; // 폴더 생성
        }
        return false;
    }

    /** 위치 정보 리스너 **/
    final LocationListener gpsLocationListener = new LocationListener() {
        public void onLocationChanged(Location location) {

            String provider = location.getProvider();
            double now_longitude = location.getLongitude();
            double now_latitude = location.getLatitude();
            double now_altitude = location.getAltitude();

            latitude = now_latitude;
            longitude = now_longitude;
            altitude = now_altitude;

            LocationManager lm = (LocationManager) getSystemService(Context.LOCATION_SERVICE);

            // 권한 체크
            if (ActivityCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(getApplicationContext(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                // TODO: Consider calling
                //    ActivityCompat#requestPermissions
                // here to request the missing permissions, and then overriding
                //   public void onRequestPermissionsResult(int requestCode, String[] permissions,
                //                                          int[] grantResults)
                // to handle the case where the user grants the permission. See the documentation
                // for ActivityCompat#requestPermissions for more details.
                return;
            }
            lm.requestLocationUpdates(LocationManager.NETWORK_PROVIDER,
                    3000,
                    1,
                    gpsLocationListener);

            lm.requestLocationUpdates(LocationManager.GPS_PROVIDER,
                    3000,
                    1,
                    gpsLocationListener);
        }

        public void onStatusChanged(String provider, int status, Bundle extras) {
        }

        public void onProviderEnabled(String provider) {
        }

        public void onProviderDisabled(String provider) {
        }
    };
}

File: app/src/main/java/kr/ac/hansung/demap/ui/main/MainActivity.kt
package kr.ac.hansung.demap.ui.main

import android.content.Context
import android.content.Intent
import android.graphics.PointF
import android.os.Bundle
import android.text.Html
import android.view.MenuItem
import android.view.View
import android.widget.ImageView
import android.widget.SearchView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.bottomnavigation.BottomNavigationView
import com.google.android.material.navigation.NavigationView
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.firestore.FirebaseFirestore
import com.naver.maps.geometry.LatLng
import com.naver.maps.map.MapFragment
import com.naver.maps.map.NaverMap
import com.naver.maps.map.OnMapReadyCallback
import com.naver.maps.map.overlay.InfoWindow
import com.naver.maps.map.overlay.Marker
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.activity_main_content.*
import kotlinx.android.synthetic.main.nav_header_main.*
import kotlinx.android.synthetic.main.nav_header_main.view.*
import kotlinx.android.synthetic.main.toolbar_main.*
import kr.ac.hansung.demap.*
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.URL
import java.net.URLEncoder
import javax.net.ssl.HttpsURLConnection


class MainActivity : AppCompatActivity(), OnMapReadyCallback,
    NavigationView.OnNavigationItemSelectedListener,
    BottomNavigationView.OnNavigationItemSelectedListener {

    var auth: FirebaseAuth? = null
    var firestore: FirebaseFirestore? = null

    var nickname: String? = null

    private var adapter // FolderList 어댑터
            : MySearchNaverRecyclerAdapter? = null


    private var br: BufferedReader? = null
    private var searchResult: StringBuilder? = null

    var datas: String? = null
    var array: Array<String?>? = null
    var title: Array<String?>? = null
    var roadaddress: Array<String?>? = null
    var mapx: IntArray? = null;
    var mapy: IntArray? = null;
    //var category: Array<String>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        //Initiate
        auth = FirebaseAuth.getInstance()
        firestore = FirebaseFirestore.getInstance()

        setSupportActionBar(toolbar_main) // 툴바를 액티비티의 앱바로 지정
        supportActionBar?.let {
            it.setBackgroundDrawable(getDrawable(R.color.colorWhite))
            it.setDisplayHomeAsUpEnabled(true) // 드로어를 꺼낼 홈 버튼 활성화
            it.setHomeAsUpIndicator(R.drawable.ic_menu_black_24dp) // 홈버튼 이미지 변경
            it.setDisplayShowTitleEnabled(false) // 툴바에 타이틀 안보이게
        }

        window.statusBarColor = resources.getColor(R.color.colorWhite, theme)

        // 로그인한 유저 닉네임 받아오기
        firestore?.collection("users")?.document(auth?.currentUser?.uid!!)?.get()?.addOnSuccessListener { documentSnapshot ->
            nickname = documentSnapshot["nickName"].toString();
            main_nav_header.tv_nickname.text = nickname
            main_nav_header.tv_email.text = auth?.currentUser?.email
        }

        //navigationListener
        main_nav.setNavigationItemSelectedListener(this)

        // bottom navigation
        bottom_nav.setOnNavigationItemSelectedListener(this)

        // naver map 객체 가져오기
        val mapFragment = supportFragmentManager.findFragmentById(R.id.map_fragment) as MapFragment?
            ?: MapFragment.newInstance().also {
                supportFragmentManager.beginTransaction().add(R.id.map_fragment, it).commit()
            }
        mapFragment.getMapAsync(this)

        // 검색어 가져오기
        //status1 = (TextView)findViewById(R.id.textview_searchresult_name); //파싱된 결과를 보자
//status2 = (TextView)findViewById(R.id.textview_seatchresult_address);
// 검색어 가져오기
        val searchView =
            findViewById<SearchView>(R.id.sv_searchPlace)
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(searchword: String): Boolean { // 검색 버튼이 눌러졌을 때 이벤트 처리
                println("검색 처리됨 : $searchword")
                //searchForFolderName(keyword);
                searchForNaverAPI(searchword)
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean { // 검색어가 변경되었을 때 이벤트 처리
                return false
            }
        })
        //System.out.println(query);

        //System.out.println(query);
        val recyclerView : RecyclerView? =
            findViewById<RecyclerView>(R.id.listView_search_result_list)
        if (recyclerView != null) {
            recyclerView.setHasFixedSize(false)
        }
        if (recyclerView != null) {
            recyclerView.layoutManager = LinearLayoutManager(this)
        }
        adapter = MySearchNaverRecyclerAdapter()
        if (recyclerView != null) {
            recyclerView.adapter = adapter
        }
    }

    fun createFolder() {
        var intent = Intent(this, CreateFolderActivity::class.java)
        startActivity(intent)
    }

    fun viewFolderList() {
        var intent = Intent(this, FolderListActivity::class.java)
        startActivity(intent)
    }

    fun viewMyFolderList() {
        var intent = Intent(this, MyfolderViewActivity::class.java)
        startActivity(intent)
    }

    fun viewSearchHotPlace() {
        var intent = Intent(this, SearchNaverActivity::class.java)
        startActivity(intent)
    }

/*

    // searchview
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        sv_searchPlace.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                // 검색어가 변경되었을 때 이벤트 처리
                return false
            }

            override fun onQueryTextChange(keyword: String): Boolean {
                Toast.makeText(this@MainActivity, keyword ,Toast.LENGTH_SHORT).show()
                return true
            }
        })
        return true
    }

*/
// 네이버 지역 검색 api 연동 및 데이터 가져오기
open fun searchForNaverAPI(query: String?): Unit {
    val clientId = "F29Q2vNcHyw0fOQwkzbO" //애플리케이션 클라이언트 아이디값";
    val clientSecret = "5dNDcpf9qo" //애플리케이션 클라이언트 시크릿값";
    val display = 5 // 보여지는 검색결과의 수
    // 네트워크 연결은 Thread 생성 필요
    object : Thread() {
        override fun run() {
            try {
                val searchword : String? = URLEncoder.encode(query, "UTF-8")
                println("검색어 utf-8 : $searchword")
                val url = URL(
                    "https://openapi.naver.com/v1/search/local?" //+ "key=" + clientSecret
                            + "&query=" + searchword //여기는 쿼리를 넣으세요(검색어)
                            + "&target=local&start=1&display=" + display
                )
                println("검색 url : $url")
                val con =
                    url.openConnection() as HttpsURLConnection
                con.requestMethod = "GET"
                con.setRequestProperty("X-Naver-Client-Id", clientId)
                con.setRequestProperty("X-Naver-Client-Secret", clientSecret)
                con.connect()
                val responseCode = con.responseCode
                br = if (responseCode == 200) { // 정상 호출
                    BufferedReader(InputStreamReader(con.inputStream))
                } else { // 에러 발생
                    BufferedReader(InputStreamReader(con.errorStream))
                }
                searchResult = java.lang.StringBuilder()
                println("207라인 정상")
                var inputLine: String = br!!.readLine()
                println("209라인 정상" + inputLine)
                while (inputLine != null) {
                    // 여기에서 FATAL 예외 발생 중
                    searchResult!!.append(inputLine + "\n")
                    inputLine = br!!.readLine()
                }
                println("212라인 정상")
                br!!.close()
                con.disconnect()
                br = if (responseCode == 200) { // 정상 호출
                    BufferedReader(InputStreamReader(con.inputStream))
                } else { // 에러 발생
                    BufferedReader(InputStreamReader(con.errorStream))
                }
                // 데이터 파싱
                datas = searchResult.toString()
                println("검색 datas : $datas")
                array = datas!!.split("\"").toTypedArray()
                title = arrayOfNulls(display)
                roadaddress = arrayOfNulls(display)
                mapx = IntArray(display)
                mapy = IntArray(display)
                //category = new String[display];
                var k = 0
                for (i in array!!.indices) {
                    if (array!![i] == "title") {
                        title!![k] = Html.fromHtml(array!![i + 2]).toString()
                        println(array!![i + 2])
                    }
                    if (array!![i] == "roadAddress") roadaddress!![k] =
                        Html.fromHtml(array!![i + 2]).toString()
                    if (array!![i] == "mapx") mapx!![k] = array!![i + 2]?.toInt()!!
                    if (array!![i] == "mapy") {
                        mapy!![k] = array!![i + 2]?.toInt()!!
                        k++
                    }
                    /*if (array[i].equals("category")) {
                            category[k] = Html.fromHtml(array[i + 2]).toString();
                            k++;
                        }*/
                }
                //System.out.println(array);
                println(
                    roadaddress!![0] + roadaddress!![1] + roadaddress!![2] + roadaddress!![3] + roadaddress!![4]
                )
                //System.out.println(category[0]+category[1]+category[2]+category[3]+category[4]);
                println(
                    "x 좌표 : " + mapx!![0] + mapx!![1] + mapx!![2] + mapx!![3] + mapx!![4]
                )
                println(
                    "y 좌표 : " + mapy!![0] + mapy!![1] + mapy!![2] + mapy!![3] + mapy!![4]
                )
                adapter!!.addItems(title, roadaddress,  /*category,*/mapx, mapy)
                adapter!!.notifyDataSetChanged()
                // title[0], link[0], bloggername[0] 등 인덱스 값에 맞게 검색결과를 변수화하였다.
            } catch (e: Exception) { //status1.setText("에러가..났습니다...");
                println("에러 발생 : $e")
            }
        }
    }.start()
}


    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId){
            android.R.id.home->{
                drawerlayout_main.openDrawer(GravityCompat.START) // 네비게이션 드로어 열기
            }
        }
        return super.onOptionsItemSelected(item)
    }


    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        when(item.itemId){
            R.id.search_folder_menu-> viewFolderList() //Toast.makeText(this,"폴더검색 clicked",Toast.LENGTH_SHORT).show()
            R.id.my_folder_menu-> viewMyFolderList() //Toast.makeText(this,"마이폴더 clicked",Toast.LENGTH_SHORT).show()
            R.id.hotplace_menu-> viewSearchHotPlace()
            R.id.history_menu-> Toast.makeText(this,"히스토리 clicked",Toast.LENGTH_SHORT).show()
            R.id.setting_menu-> Toast.makeText(this,"설정 clicked",Toast.LENGTH_SHORT).show()
            R.id.service_menu-> Toast.makeText(this,"고객센터 clicked",Toast.LENGTH_SHORT).show()
            R.id.logout_menu-> Toast.makeText(this,"로그아웃 clicked",Toast.LENGTH_SHORT).show()

            R.id.history_bottom_nav-> Toast.makeText(this,"바텀히스토리 clicked",Toast.LENGTH_SHORT).show()
            R.id.search_folder_bottom_nav-> viewFolderList()
            R.id.my_folder_bottom_nav-> createFolder()
        }
        return false
    }

    // 드로어가 나와있을 때 뒤로가기 버튼 누를 경우 드로어 닫기
    override fun onBackPressed() {
        if (drawerlayout_main.isDrawerOpen(GravityCompat.START)) {
            drawerlayout_main.closeDrawers()
        } else {
            super.onBackPressed()
        }
    }


    // 좌표 눌렀을 때 커스텀 window adapter
    private class InfoWindowAdapter(private val context: Context) : InfoWindow.ViewAdapter() {
        private var rootView: View? = null
        private var icon: ImageView? = null
        private var text: TextView? = null

        override fun getView(infoWindow: InfoWindow): View {
            val view = rootView ?: View.inflate(context, R.layout.view_custom_info_window, null).also { rootView = it }
            val icon = icon ?: view.findViewById<ImageView>(R.id.icon).also { icon = it }
            val text = text ?: view.findViewById<TextView>(R.id.text).also { text = it }

            val marker = infoWindow.marker
            if (marker != null) {
                icon.setImageResource(R.drawable.ic_place_black_24dp)
                text.text = marker.tag as String?
            } else {
                icon.setImageResource(R.drawable.ic_my_location_black_24dp)
                text.text = context.getString(
                    R.string.format_coord, infoWindow.position.latitude, infoWindow.position.longitude)
            }
            return view
        }
    }



    // 지도 사용하는 거 다 여기에 쓰면 되는거가타요
    override fun onMapReady(naverMap: NaverMap) {
        val infoWindow = InfoWindow().apply {
            anchor = PointF(0f, 1f)
            offsetX = resources.getDimensionPixelSize(R.dimen.custom_info_window_offset_x)
            offsetY = resources.getDimensionPixelSize(R.dimen.custom_info_window_offset_y)
            adapter = InfoWindowAdapter(this@MainActivity)
            setOnClickListener {
                close()
                true
            }
        }


        val marker = Marker()
        marker.position = LatLng(37.5670135, 126.9783740)
        marker.map = naverMap

        // 지도 아무데나 눌렀을 때
        naverMap.setOnMapClickListener { _, coord ->
            Toast.makeText(this, "${coord}", Toast.LENGTH_LONG).show()
            var now : LatLng? = coord

            if(now != marker.position) {

                marker.map = null

                marker.position = coord
                marker.map = naverMap

            }
            /*
            infoWindow.position = coord
            infoWindow.open(naverMap)
            */
        }
    }
}
